<a href="https://colab.research.google.com/github/jesusvasquez-tec/analitica-empresarial-integrada/blob/main/Semana6/Vasquez_LAB_D6_AEI_2026_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ANALÍTICA EMPRESARIAL INTEGRADA
## LABORATORIO DIRIGIDO N.° 06

### DISEÑO Y ANÁLISIS DE UN EXPERIMENTO A/B

**Semana 6 · TECSUP 2026-II**  
**Docente:** Pilar Rocío Sayán Mejía  
**Carrera:** Big Data y Ciencia de Datos  
**Plataforma de trabajo:** Google Colab


**Alumno:** Jesus Vasquez Espinoza
### Pregunta orientadora

**¿Qué evidencia estadística y práctica necesitamos antes de implementar una variante B?**

> Este notebook acompaña el trabajo desarrollado durante la clase. No es un informe ni un proyecto para exponer. Las celdas alternan explicación breve, ejecución guiada, interpretación y retos de cierre.


# I. Elemento de la capacidad terminal de la semana

Diseña y analiza un experimento A/B identificando el factor, sus niveles, el control, el tratamiento, la unidad experimental y la variable respuesta. Formula las hipótesis estadísticas, establece la métrica primaria, los guardrails, el nivel de significación, la potencia y el efecto mínimo detectable; finalmente, integra la evidencia estadística con la relevancia práctica para sustentar una decisión.


# II. Seguridad y condiciones de trabajo

- Utilice únicamente Google Colab y las fuentes indicadas por la docente.
- No modifique ni fabrique resultados para forzar una conclusión.
- Ejecute las celdas en el orden propuesto y documente cualquier error o dato ausente.
- No comparta claves, credenciales ni información personal en el notebook.
- Mantenga la estación de trabajo ordenada y no instale software no autorizado.


# III. Fundamento teórico
Un A/B test compara un control A con una variante B mediante asignación aleatoria. La unidad experimental de este caso es el usuario, que conserva su variante durante toda la observación. La aleatorización favorece la comparabilidad en expectativa; no garantiza grupos idénticos.

La respuesta es binaria: compra dentro de los siete días posteriores a la asignación. Todos los usuarios completan esa ventana antes del análisis. La métrica primaria es la proporción de usuarios que compran. Se estudia la diferencia de dos proporciones, no ANOVA.

El MDE es el efecto utilizado para planificar la potencia. El umbral empresarial es la mejora que justificaría interés práctico. Aquí son distintos: MDE de 1,5 pp y umbral empresarial de 1,0 pp. La muestra se dimensiona para detectar el MDE frente a cero, no para garantizar evidencia de superar el umbral empresarial.

Se usa una aproximación normal de Wald para la diferencia. El contraste unilateral al 5 % y el límite inferior unilateral del 95 % utilizan el mismo error estándar. También se muestra un IC bilateral del 95 % como descripción de incertidumbre; no se exige que este último excluya cero para reproducir el contraste unilateral.

La recomendación académica requiere evidencia de superar el umbral empresarial y de mantener el guardrail dentro del margen predefinido. Son criterios más exigentes que obtener p < 0,05 frente a cero. Un resultado no concluyente no demuestra igualdad.


# IV. Normas y buenas prácticas empleadas

- Prerregistro lógico: definir hipótesis, métrica, MDE y regla de decisión antes de observar los resultados.
- Aleatorización: evitar que la asignación dependa de características de las unidades.
- Trazabilidad: conservar la fuente, las transformaciones y los parámetros utilizados.
- Reproducibilidad: ejecutar el análisis mediante código verificable.
- No realizar *peeking*: no detener o prolongar el experimento solo porque el resultado momentáneo conviene.
- No contaminar grupos: una unidad debe recibir únicamente la variante asignada.
- Distinguir asociación, evidencia experimental y decisión práctica.


# V. Recursos
- Computadora, Google Colab y notebook LAB-D6-AEI-PSAYAN-2026-02.ipynb.
- Librerías numpy, pandas, scipy, statsmodels y plotly.
- Caso ficticio NexaCommerce con datos simulados generados en el propio notebook. No requiere descargar un dataset externo.

## Naturaleza de los datos utilizados
Los registros son sintéticos y reproducibles mediante una semilla fija. Representan usuarios de un checkout; no proceden de una empresa real. Las probabilidades del simulador son supuestos académicos, no evidencia observada. No cambie la semilla para buscar un resultado favorable. En un estudio real, los datos procederían de registros del experimento y el plan se documentaría antes de recogerlos.


# VI. Metodología para el desarrollo de la sesión

El laboratorio se desarrolla mediante una secuencia guiada de cinco momentos:

1. **Comprender:** recuperar el vocabulario esencial.
2. **Representar:** traducir el caso a la estructura de un diseño experimental.
3. **Formular:** establecer las hipótesis y la regla de decisión antes de observar resultados.
4. **Ejecutar e interpretar:** correr el código por etapas y explicar cada salida.
5. **Decidir y transferir:** integrar evidencia estadística, umbral empresarial y guardrails; luego resolver retos breves.

No se solicita exposición, informe ni trabajo adicional. La evidencia de aprendizaje se construye durante la clase mediante respuestas breves en las pausas de análisis.


# VII. Procedimiento

## Actividad 1 — Activación conceptual

**Propósito:** reconocer los componentes que deben definirse antes de ejecutar un experimento.

Complete la tabla con sus propias palabras. La pregunta de apoyo permite verificar si la definición es suficiente.

| Concepto | Definición con sus propias palabras | Pregunta de apoyo |
|---|---|---|
| Experimento A/B | Es una comparación controlada y aleatorizada entre dos variantes para estimar el efecto causal de un cambio, no solo una observación pasiva de grupos. | ¿Qué diferencia existe entre comparar dos grupos y experimentar? |
| Control | Es la versión de referencia o situación actual del proceso, contra la cual se evalúa el impacto de la nueva variante. | ¿Qué representa el punto de referencia? |
| Tratamiento | Es la nueva variante o modificación específica que se desea evaluar frente al grupo de control. | ¿Qué cambio recibe este grupo? |
| Unidad experimental | Es el elemento individual (en este caso, el usuario) al que se le asigna aleatoriamente una variante y sobre el cual se mide el resultado. | ¿Sobre quién o qué se aplica la variante? |
| Variable respuesta | Es el resultado medible (generalmente binario) que se registra en cada unidad experimental tras la exposición a la variante (ej. compra_7d). | ¿Qué resultado se observa en cada unidad? |
| Métrica primaria | Es el indicador principal de éxito, definido antes de ver los datos, que responde directamente a la decisión de negocio (ej. tasa de conversión). | ¿Cuál será el criterio principal de éxito? |
| Guardrail | Es una métrica de seguridad o secundaria que debe monitorearse para asegurar que no se deteriora más allá de un margen permitido, aunque la métrica primaria mejore. | ¿Qué no debería deteriorarse? |
| MDE | Es la magnitud mínima del efecto que el experimento está diseñado para detectar con una potencia estadística dada; se usa para calcular el tamaño de la muestra. | ¿Qué efecto se utiliza para planificar la potencia? |
| Potencia | Es la probabilidad (ej. 80 %) de que el diseño del experimento detecte correctamente el MDE si este efecto existe realmente, minimizando el error Tipo II. | ¿Qué capacidad tiene el diseño para detectar el MDE? |
| Peeking | Es la práctica incorrecta de revisar los resultados continuamente y detener el experimento prematuramente al ver significancia, lo que infla el error Tipo I. | ¿Por qué no se debe decidir mirando resultados parciales? |

**Pausa de verificación:** explique a un compañero por qué el MDE no debe elegirse después de conocer la diferencia observada.


## Actividad 2 — Interpretación y representación metodológica
NexaCommerce, empresa ficticia, compara el checkout actual de cuatro pasos con uno simplificado de dos pasos. Cada usuario recibe una sola variante. Se mide compra e incidencia de pago dentro de siete días; incluso quienes no compran permanecen en el denominador.

### 2.1 Interpretación
Complete antes de ejecutar: Determinar si el checkout simplificado de dos pasos (B) aumenta la tasa de conversión de compra respecto al checkout actual de cuatro pasos (A), manteniendo protegido el guardrail de incidencia de pagos.
.

### 2.2 Representación del diseño
| Elemento | Definición en este experimento |
|---|---|
| Factor y niveles | Checkout: A actual; B simplificado |
| Unidad experimental | Usuario único con asignación persistente |
| Respuesta primaria | compra_7d: 1 si compró dentro de siete días; 0 si no |
| Métrica primaria | Usuarios que compraron / todos los usuarios asignados |
| Guardrail | incidencia_pago_7d: al menos una incidencia por usuario en siete días |
| Denominador del guardrail | Todos los usuarios asignados, compren o no |
| Margen de deterioro | B−A debe ser inferior a +0,5 pp en incidencias |
| Baseline histórico supuesto | Conversión 11,8 %; incidencias 1,8 % |
| MDE de conversión | +1,5 pp para planificar potencia de 80 % frente a cero |
| Umbral empresarial | Mejora de conversión superior a +1,0 pp |
| Horizonte | Reclutamiento planificado de 14 días y 7 días de maduración |
| Muestra | Se calcula antes de simular las respuestas |

### 2.3 Hipótesis y reglas previas
Conversión: H0: pB−pA ≤ 0; H1: pB−pA > 0. Alfa = 0,05 unilateral.

Relevancia empresarial: exigir límite inferior unilateral del 95 % mayor que 0,010. Esto evalúa un umbral distinto de cero y no se deduce solo de la estimación puntual.

Guardrail: H0: qB−qA ≥ 0,005; H1: qB−qA < 0,005. Exigir límite superior unilateral del 95 % menor que 0,005. No encontrar una diferencia significativa no demuestra que el guardrail esté protegido.

La recomendación combina ambos requisitos; se trata de una regla conjunta, no de escoger la prueba favorable. Los intervalos individuales no deben describirse como una región de confianza simultánea.

**Comprobación:** ¿por qué se comparan proporciones? ¿Cómo cambiarían las hipótesis si interesara cualquier diferencia, positiva o negativa?
**Respuesta:** Se comparan proporciones porque la variable respuesta (compra_7d) es binaria (compró o no compró), por lo que la métrica primaria es naturalmente una proporción: el cociente entre usuarios que compraron y el total de usuarios asignados. Si interesara detectar cualquier diferencia, positiva o negativa, las hipótesis pasarían a ser bilaterales: H₀: p_B − p_A = 0 frente a H₁: p_B − p_A ≠ 0, y el nivel de significancia α = 0,05 se repartiría en ambas colas de la distribución, en lugar de concentrarse en una sola como en el diseño unilateral actual..


## Actividad 3 — Preparación y planificación previa
### Paso 1. Preparar Google Colab
Ejecute las celdas en orden. La configuración y el cálculo de muestra aparecen antes de generar las respuestas. En Colab, la primera celda instala solo los paquetes que faltan. Si se solicita reiniciar, vuelva a ejecutar desde aquí.


In [ ]:
import importlib.util
import subprocess
import sys

paquetes = ["numpy", "pandas", "scipy", "statsmodels", "plotly"]
faltan = [p for p in paquetes if importlib.util.find_spec(p) is None]
if faltan:
    subprocess.check_call([sys.executable, "-m", "pip", "install", "-q", *faltan])
print("Dependencias disponibles.")


Dependencias disponibles.


**Interpretación del resultado:**

El entorno de ejecución está correctamente configurado. Todas las librerías necesarias para el análisis estadístico y la visualización de datos del experimento A/B (numpy, pandas, scipy, statsmodels y plotly) están instaladas y disponibles para su uso.

In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px
from scipy.stats import norm
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

pd.set_option("display.max_columns", 20)
print("Entorno listo.")


Entorno listo.


**Interpretación del resultado:**

El entorno de ejecución está correctamente configurado. Todas las librerías necesarias para el análisis estadístico y la visualización de datos del experimento A/B (numpy, pandas, scipy, statsmodels y plotly) están cargadas y listas para su uso.

### Paso 2. Fijar los parámetros antes de observar
Los baselines son supuestos históricos del caso ficticio, no estimaciones obtenidas de los grupos que analizaremos. El margen del guardrail se fija por criterio de negocio. Se calcula muestra para conversión y, por aproximación normal, para no inferioridad del guardrail suponiendo diferencia verdadera cero. Se toma el mayor tamaño. Las potencias se refieren a cada objetivo de diseño; no garantizan 80 % para la regla conjunta ni para superar el umbral empresarial.


In [ ]:
ALPHA = 0.05
POTENCIA = 0.80
BASELINE = 0.118
MDE = 0.015
UMBRAL_NEGOCIO = 0.010
BASE_GUARDRAIL = 0.018
MARGEN_GUARDRAIL = 0.005
SEMILLA = 20260921
DIAS_RECLUTAMIENTO = 14
VENTANA_DIAS = 7

assert 0 < BASELINE < BASELINE + MDE < 1
assert 0 < UMBRAL_NEGOCIO < MDE
analysis = NormalIndPower()
es = proportion_effectsize(BASELINE + MDE, BASELINE)
n_conversion = int(np.ceil(analysis.solve_power(
    effect_size=es, alpha=ALPHA, power=POTENCIA,
    ratio=1, alternative="larger")))
z_alpha = norm.ppf(1 - ALPHA)
z_power = norm.ppf(POTENCIA)
n_guardrail = int(np.ceil(
    2 * BASE_GUARDRAIL * (1 - BASE_GUARDRAIL)
    * (z_alpha + z_power)**2 / MARGEN_GUARDRAIL**2))
N_POR_GRUPO = max(n_conversion, n_guardrail)
print("Muestra por grupo para conversión:", n_conversion)
print("Muestra aproximada por grupo para guardrail:", n_guardrail)
print("Muestra planificada por grupo:", N_POR_GRUPO)
print("Analizar una sola vez al completar muestra y seguimiento.")


Muestra por grupo para conversión: 6027
Muestra aproximada por grupo para guardrail: 8743
Muestra planificada por grupo: 8743
Analizar una sola vez al completar muestra y seguimiento.


**Interpretación del resultado:**

El cálculo indica que se requieren 6.027 usuarios por grupo para detectar el MDE de conversión de 1,5 pp con una potencia del 80 %. Sin embargo, para garantizar que el guardrail (incidencia de pago) no se deteriora más allá del margen permitido de 0,5 pp, se necesita una muestra mayor de 8.743 usuarios por grupo. Por lo tanto, la muestra planificada final se fija en 8.743 usuarios por grupo, ya que el requisito del guardrail es el factor más restrictivo y determina el tamaño muestral necesario para asegurar la validez estadística de ambas métricas simultáneamente.

### Pausa de análisis 1
1. ¿Por qué el baseline debe establecerse antes de ver el resultado?

    El baseline debe establecerse antes para evitar sesgos de confirmación y mantener la validez estadística del cálculo de potencia; si se ajustara tras observar los datos, el dimensionamiento de la muestra perdería su fundamento y se podría inflar la tasa de error Tipo I.

2. ¿Por qué la muestra final considera también el guardrail?

    La muestra final considera el guardrail porque el diseño debe garantizar potencia suficiente para ambos objetivos simultáneamente (detectar mejora y proteger el guardrail). El código evidencia que el requisito del guardrail es más restrictivo (8743 usuarios) que el de conversión (6027 usuarios), por lo que este determina el tamaño final.

3. ¿La potencia calculada garantiza demostrar una mejora superior a 1,0 pp?
**Respuestas:**

  No, la potencia calculada no garantiza demostrar una mejora superior a 1,0 pp. La potencia del 80 % se dimensionó para detectar el MDE de 1,5 pp frente a cero; el umbral empresarial de 1,0 pp es un criterio de relevancia práctica distinto que, con esta muestra, no tiene garantizada la misma potencia estadística.



### Paso 3. Generar y validar el conjunto académico
El simulador asigna aleatoriamente el mismo número de usuarios a A y B. Genera una fila por usuario y completa siete días de observación para todos. Supone conversiones de 11,8 % y 13,8 % e incidencias de 1,8 % en ambos grupos. Son parámetros del generador, no resultados que deban recuperarse exactamente. Las respuestas se simulan independientes para simplificar el ejercicio; una empresa real puede presentar dependencias y estacionalidad no representadas aquí.


In [ ]:
rng = np.random.default_rng(SEMILLA)
total = 2 * N_POR_GRUPO
grupos = rng.permutation(np.repeat(["A_Control", "B_Tratamiento"], N_POR_GRUPO))
prob_compra = np.where(grupos == "A_Control", 0.118, 0.138)
df = pd.DataFrame({
    "usuario_id": np.arange(1, total + 1),
    "grupo": grupos,
    "dia_asignacion": rng.integers(1, DIAS_RECLUTAMIENTO + 1, total),
    "dias_observados": np.full(total, VENTANA_DIAS),
    "compra_7d": rng.binomial(1, prob_compra),
    "incidencia_pago_7d": rng.binomial(1, BASE_GUARDRAIL, total)
})
print("DATOS SIMULADOS — NexaCommerce")
print("Dimensión:", df.shape)
display(df.head())


DATOS SIMULADOS — NexaCommerce
Dimensión: (17486, 6)


,usuario_id,grupo,dia_asignacion,dias_observados,compra_7d,incidencia_pago_7d
0,1,A_Control,8,7,0,0
1,2,B_Tratamiento,5,7,0,0
2,3,A_Control,12,7,0,0
3,4,A_Control,9,7,0,0
4,5,B_Tratamiento,12,7,0,0


**Interpretación del resultado:**

El código ha generado exitosamente el conjunto de datos sintético para el experimento, cumpliendo con el tamaño de muestra planificado de 8.743 usuarios por grupo (17.486 registros en total). La estructura del DataFrame incluye todas las variables clave definidas en el diseño: un identificador único por usuario, la asignación aleatoria al grupo (Control o Tratamiento), la ventana completa de observación (7 días) y las variables respuesta binarias (compra_7d e incidencia_pago_7d), quedando los datos listos para la validación estructural y el análisis estadístico posterior.

In [ ]:
requeridas = {"usuario_id", "grupo", "dias_observados",
              "compra_7d", "incidencia_pago_7d"}
assert requeridas.issubset(df.columns), "Faltan columnas"
assert not df[list(requeridas)].isna().any().any(), "Hay datos ausentes"
assert df["usuario_id"].is_unique, "Usuarios repetidos o contaminados"
assert set(df["grupo"]) == {"A_Control", "B_Tratamiento"}
assert (df["dias_observados"] == VENTANA_DIAS).all()
for col in ["compra_7d", "incidencia_pago_7d"]:
    assert df[col].isin([0, 1]).all(), f"{col} no es binaria"
ab = df.copy()
tam_grupos = ab.groupby("grupo").size()
assert (tam_grupos == N_POR_GRUPO).all()
display(tam_grupos.to_frame("usuarios"))
print("Calidad estructural y seguimiento completos.")


,usuarios
grupo,
A_Control,8743
B_Tratamiento,8743


Calidad estructural y seguimiento completos.


**Interpretación del resultado:**

El dataset supera todas las validaciones de calidad estructural definidas en el diseño experimental. Ambos grupos (A_Control y B_Tratamiento) cuentan con exactamente 8.743 usuarios, cumpliendo el tamaño muestral planificado sin desviaciones. No se detectaron datos ausentes, usuarios duplicados ni contaminación entre grupos; además, las variables respuesta (compra_7d e incidencia_pago_7d) conservan su naturaleza binaria y todos los registros completan la ventana de observación de 7 días. Esto garantiza que la comparación posterior entre A y B se realizará sobre datos íntegros, comparables y alineados con el prerregistro del experimento.

### Pausa de análisis 2
1. Registre el tamaño de cada grupo. ¿Cumple el plan?

    Tamaño de grupo: Ambos grupos (A_Control y B_Tratamiento) tienen exactamente 8.743 usuarios, cumpliendo estrictamente con el tamaño muestral planificado en el diseño.

2. ¿Qué ocurriría si un usuario apareciera en ambas variantes?

    Contaminación: Si un usuario apareciera en ambas variantes, se rompería la independencia de las observaciones y la comparabilidad de los grupos, invalidando la inferencia causal del experimento.

3. ¿Por qué no debe excluirse a quienes no compraron?

    Denominador constante: No deben excluirse quienes no compraron porque la métrica primaria es una proporción cuyo denominador debe ser el total de usuarios asignados; excluirlos inflaría artificialmente la tasa de conversión y sesgaría el resultado.

4. ¿Por qué es necesario completar los siete días antes de analizar?

    **Respuestas:** Ventana de observación: Es necesario completar los siete días para garantizar que todas las unidades tengan la misma oportunidad temporal de realizar la acción, evitando así sesgos por tiempos de exposición desiguales.



## Actividad 4 — MDE, potencia y sensibilidad de muestra
### Paso 4. Recuperar el diseño fijado
Revise los parámetros sin recalibrarlos con el resultado observado. El MDE sirve para dimensionar la muestra; el umbral empresarial se utiliza para valorar la mejora. En este caso, las dos cantidades difieren deliberadamente.


In [ ]:
display(pd.DataFrame({
    "parametro": ["Baseline", "MDE", "Umbral empresarial",
                  "Margen guardrail", "Alfa", "Potencia"],
    "valor": [BASELINE, MDE, UMBRAL_NEGOCIO,
              MARGEN_GUARDRAIL, ALPHA, POTENCIA]
}))
print("Muestra final por grupo:", N_POR_GRUPO)


,parametro,valor
0,Baseline,0.118
1,MDE,0.015
2,Umbral empresarial,0.010
3,Margen guardrail,0.005
4,Alfa,0.050
5,Potencia,0.800


Muestra final por grupo: 8743


**Interpretación del resultado:**

La salida muestra la recuperación de los parámetros de diseño fijados antes de observar los datos, garantizando el prerregistro lógico del experimento. Se evidencia la distinción metodológica clave entre el MDE (1,5 pp), utilizado para calcular la potencia estadística, y el umbral empresarial (1,0 pp), que define la relevancia práctica de la decisión. Además, se confirma que la muestra final de 8.743 usuarios por grupo está determinada por el requisito más restrictivo del guardrail, asegurando que el diseño tenga suficiente capacidad para proteger la métrica de seguridad.

### Paso 5. Explorar la sensibilidad al MDE
Compare MDE de 0,5, 1,5 y 2,5 pp. Esta exploración es didáctica y no cambia el plan ni la decisión del caso ya fijado.
**Predicción:** ¿qué efecto necesitará más muestra y por qué?


In [ ]:
filas = []
for mde in [0.005, 0.015, 0.025]:
    es = proportion_effectsize(BASELINE + mde, BASELINE)
    n = analysis.solve_power(effect_size=es, alpha=ALPHA,
                             power=POTENCIA, alternative="larger")
    filas.append({"MDE_pp": mde * 100,
                  "n_conversion": int(np.ceil(n)),
                  "n_con_guardrail": max(int(np.ceil(n)), n_guardrail)})
sensibilidad_n = pd.DataFrame(filas)
display(sensibilidad_n)
fig = px.line(sensibilidad_n, x="MDE_pp",
              y=["n_conversion", "n_con_guardrail"], markers=True,
              title="Muestra por grupo según el MDE",
              labels={"value": "Usuarios por grupo", "MDE_pp": "MDE (pp)"})
fig.show()


,MDE_pp,n_conversion,n_con_guardrail
0,0.5,52413,52413
1,1.5,6027,8743
2,2.5,2241,8743


**Interpretación del resultado:**

La tabla y el gráfico evidencian la relación inversa entre el MDE y el tamaño de muestra requerido para la métrica primaria: a menor efecto a detectar (0,5 pp), mayor es la muestra necesaria (52.413 usuarios por grupo). Sin embargo, se observa que el requisito del guardrail actúa como un piso mínimo de 8.743 usuarios. Cuando el MDE es de 1,5 pp o 2,5 pp, el cálculo para conversión (6.027 y 2.241, respectivamente) resulta inferior al del guardrail, por lo que este último termina determinando el tamaño muestral final del experimento.

### Pausa de análisis 3
1. Explique por qué reducir el MDE aumenta la muestra para conversión.

    Reducir el MDE aumenta la muestra porque detectar un efecto más pequeño exige mayor precisión estadística para distinguirlo del ruido aleatorio. El código evidencia esta relación inversa al mostrar que, manteniendo la potencia del 80 %, bajar el MDE de 1,5 pp a 0,5 pp incrementa la muestra requerida para conversión de 6.027 a 52.413 usuarios por grupo.

2. ¿Por qué el guardrail puede determinar la muestra final?

    El guardrail puede determinar la muestra final porque el diseño experimental debe garantizar potencia suficiente tanto para la métrica primaria como para la métrica de seguridad. En este caso, el cálculo de no inferioridad para el guardrail exigió 8.743 usuarios, superando los 6.027 de la conversión; por ello, la función max() del código impone el valor más restrictivo para proteger ambos objetivos simultáneamente.

3. ¿Sería válido aumentar el MDE después de ver los resultados para declarar éxito?

    **Respuestas:** No es válido aumentar el MDE después de ver los resultados, ya que es un parámetro de diseño que debe fijarse antes de observar los datos para mantener la validez de la potencia. Modificarlo a posteriori para declarar éxito invalida el prerregistro del experimento, introduce sesgo de confirmación y constituye una práctica de p-hacking.



## Actividad 5 — Análisis estadístico de la comparación A/B

**Propósito:** calcular el efecto observado y evaluar la hipótesis mediante una prueba para dos proporciones.

### Paso 6. Obtener tasas y tamaño del efecto observado

Antes de ejecutar, recuerde:

- Diferencia absoluta: $p_B-p_A$.
- Lift relativo: $(p_B-p_A)/p_A$.
- La diferencia se expresa en puntos porcentuales; el lift expresa un cambio relativo respecto de A.


In [ ]:
tabla = ab.groupby("grupo")["compra_7d"].agg(["sum", "count", "mean"])
tabla.columns = ["compras", "n", "tasa_conversion"]
x_A, n_A = int(tabla.loc["A_Control", "compras"]), int(tabla.loc["A_Control", "n"])
x_B, n_B = int(tabla.loc["B_Tratamiento", "compras"]), int(tabla.loc["B_Tratamiento", "n"])
p_A, p_B = x_A / n_A, x_B / n_B
efecto = p_B - p_A
lift_rel = efecto / p_A if p_A > 0 else np.nan
display(tabla)
print(f"Diferencia B−A: {efecto*100:.3f} pp")
print(f"Lift relativo: {lift_rel:.2%}")


,compras,n,tasa_conversion
grupo,,,
A_Control,996,8743,0.11392
B_Tratamiento,1236,8743,0.14137


Diferencia B−A: 2.745 pp
Lift relativo: 24.10%


**Interpretación del resultado:**

La variante B (checkout simplificado) alcanzó una tasa de conversión del 14,14 %, frente al 11,39 % del grupo de control A. Esto representa una mejora absoluta de 2,745 puntos porcentuales (pp) y un incremento relativo (lift) del 24,10 %. Estos valores describen el comportamiento observado en la muestra, indicando una tendencia positiva a favor del tratamiento; sin embargo, son únicamente métricas descriptivas y aún no constituyen evidencia estadística de un efecto real hasta que se aplique la prueba de hipótesis correspondiente.

### Pausa de análisis 4

Complete sin utilizar todavía el p-valor:

> La tasa de B es 14,14 % y la tasa de A es 11,39 %. La diferencia observada es 2,745 puntos porcentuales, equivalente a un lift relativo de 24,10 %. Estos valores describen la muestra, pero todavía no permiten afirmar que la diferencia sea estadísticamente significativa o que no se deba únicamente a la variabilidad aleatoria del muestreo.


### Paso 7. Contraste e intervalos coherentes
Usamos la aproximación normal de Wald con error estándar no combinado. Se comprueba que haya al menos diez compras y diez no compras en cada grupo; esta regla práctica no sustituye la revisión del diseño.

Para la prueba unilateral frente a cero, p < 0,05 equivale a límite inferior unilateral del 95 % > 0 con el mismo método. El IC bilateral del 95 % se presenta aparte. La función también permite contrastar un umbral distinto de cero. El p-valor no es la probabilidad de que H0 sea verdadera.


In [ ]:
def diferencia_wald(x_b, n_b, x_a, n_a, umbral=0.0):
    if min(x_b, n_b-x_b, x_a, n_a-x_a) < 10:
        raise ValueError("Frecuencias pequeñas: revisar método antes de continuar")
    pb, pa = x_b/n_b, x_a/n_a
    d = pb-pa
    se = np.sqrt(pb*(1-pb)/n_b + pa*(1-pa)/n_a)
    z = (d-umbral)/se
    return {"d": d, "se": se, "z": z,
            "p_mayor": norm.sf(z), "p_menor": norm.cdf(z),
            "li_uni95": d-norm.ppf(1-ALPHA)*se,
            "ls_uni95": d+norm.ppf(1-ALPHA)*se,
            "ic95_low": d-norm.ppf(1-ALPHA/2)*se,
            "ic95_high": d+norm.ppf(1-ALPHA/2)*se}

conv = diferencia_wald(x_B, n_B, x_A, n_A)
negocio = diferencia_wald(x_B, n_B, x_A, n_A, UMBRAL_NEGOCIO)
print(f"z frente a cero: {conv['z']:.4f}")
print(f"p unilateral frente a cero: {conv['p_mayor']:.6g}")
print(f"Límite inferior unilateral 95%: {conv['li_uni95']*100:.3f} pp")
print(f"IC bilateral 95%: [{conv['ic95_low']*100:.3f}, {conv['ic95_high']*100:.3f}] pp")
print(f"p unilateral frente a +1,0 pp: {negocio['p_mayor']:.6g}")


z frente a cero: 5.4436
p unilateral frente a cero: 2.61091e-08
Límite inferior unilateral 95%: 1.916 pp
IC bilateral 95%: [1.757, 3.733] pp
p unilateral frente a +1,0 pp: 0.000269553


**Interpretación del resultado:**

* Alfa, z y p-valor: Alfa = 0,05; estadístico z = 5,4436; p-valor unilateral frente a cero = 2,61 × 10⁻⁸.

* Rechazo de H₀: Sí, se rechaza H₀. Dado que el p-valor (2,61 × 10⁻⁸) es considerablemente menor que el nivel de significancia (α = 0,05), existe evidencia estadística sólida para afirmar que la variante B (checkout simplificado) genera una tasa de conversión mayor que la variante A (checkout actual).

* Límites y umbral: Sí, el límite inferior unilateral del 95 % es de 1,916 pp, superando tanto el cero como el umbral empresarial de 1,0 pp.

* Diferencia entre intervalos: El IC bilateral del 95 % distribuye el error alfa (0,05) en ambas colas de la distribución (0,025 por cola, z ≈ 1,96), mientras que el límite unilateral del 95 % concentra todo el error alfa en una sola cola (z ≈ 1,645). Esto hace que el extremo inferior del intervalo unilateral sea más alto (menos conservador para probar una mejora direccional) que el del bilateral.

* Evidencia suficiente: No. Una estimación puntual superior a 1,0 pp (2,745 pp) no garantiza por sí sola que el efecto real supere ese umbral. Se requiere que el límite inferior del intervalo de confianza unilateral del 95 % también sea mayor a 1,0 pp (como ocurre aquí con 1,916 pp) para afirmar con suficiente evidencia estadística que la mejora es empresarialmente relevante, descontando la incertidumbre del muestreo.

In [ ]:
grafico = tabla.reset_index()
grafico["conversion_pct"] = grafico["tasa_conversion"] * 100
fig = px.bar(grafico, x="grupo", y="conversion_pct", text="conversion_pct",
             title="Conversión observada — datos simulados",
             labels={"conversion_pct": "Usuarios que compraron (%)", "grupo": "Grupo"})
fig.update_traces(texttemplate="%{text:.2f}%", textposition="outside")
fig.update_yaxes(rangemode="tozero")
fig.show()


**Interpretación del gráfico:**

La visualización de barras refleja de manera clara la diferencia descriptiva entre ambos grupos: el grupo B_Tratamiento alcanza una tasa de conversión del 14,14 %, mientras que el grupo A_Control se sitúa en el 11,39 %. El gráfico ilustra la magnitud del efecto observado en la muestra, mostrando una ventaja visual notable a favor del checkout simplificado. Sin embargo, esta representación gráfica solo describe el comportamiento muestral y no incorpora la incertidumbre estadística (intervalos de confianza ni p-valor), por lo que no permite por sí sola concluir si la diferencia es estadísticamente significativa o atribuible al azar.

### Análisis y conclusión estadística
1. Registre alfa, z y p unilateral frente a cero.

    Alfa, z y p-valor: Alfa = 0,05; estadístico z = 5,4436; p-valor unilateral frente a cero = 2,61 × 10⁻⁸.

2. ¿Se rechaza H0? Explique en el contexto de conversión del caso simulado.

    Decisión sobre H₀: Sí, se rechaza H₀. Dado que el p-valor (2,61 × 10⁻⁸) es considerablemente menor que el nivel de significancia (α = 0,05), existe evidencia estadística sólida para afirmar que la variante B (checkout simplificado) genera una tasa de conversión significativamente mayor que la variante A (checkout actual).

3. ¿El límite inferior unilateral supera cero? ¿Y el umbral de 1,0 pp?

    Límites y umbral: Sí, el límite inferior unilateral del 95 % es de 1,916 pp, superando tanto el valor cero como el umbral empresarial de 1,0 pp.

4. Explique por qué el IC bilateral del 95 % y el límite unilateral del 95 % tienen extremos distintos.

    Diferencia entre intervalos: El IC bilateral del 95 % distribuye el error alfa (0,05) en ambas colas de la distribución (0,025 por cola, z ≈ 1,96), mientras que el límite unilateral del 95 % concentra todo el error alfa en una sola cola (z ≈ 1,645). Esto hace que el extremo inferior del intervalo unilateral sea más alto (menos conservador para probar una mejora direccional) que el del bilateral.

5. ¿Superar 1,0 pp con la estimación puntual demuestra superar ese umbral con evidencia suficiente?

    **Respuestas:** Evidencia suficiente: No. Una estimación puntual superior a 1,0 pp (2,745 pp) no garantiza por sí sola que el efecto real supere ese umbral. Se requiere que el límite inferior del intervalo de confianza unilateral del 95 % también sea mayor a 1,0 pp (como ocurre aquí con 1,916 pp) para afirmar con suficiente evidencia estadística que la mejora es empresarialmente relevante, descontando la incertidumbre del muestreo.

## Actividad 6 — Del resultado estadístico a la decisión
### Paso 8. Evaluar el guardrail con incertidumbre
El guardrail es la proporción de usuarios con alguna incidencia de pago en siete días. El margen de deterioro admisible es +0,5 pp. La no inferioridad exige que el límite superior unilateral del 95 % esté por debajo de ese margen. No basta obtener p > 0,05 en una prueba de igualdad.


In [ ]:
guardrail = ab.groupby("grupo")["incidencia_pago_7d"].agg(["sum", "count", "mean"])
guardrail.columns = ["incidencias", "n", "tasa_incidencia"]
g_A = int(guardrail.loc["A_Control", "incidencias"])
g_B = int(guardrail.loc["B_Tratamiento", "incidencias"])
guard = diferencia_wald(g_B, n_B, g_A, n_A, MARGEN_GUARDRAIL)
display(guardrail)
print(f"Cambio del guardrail B−A: {guard['d']*100:.3f} pp")
print(f"Límite superior unilateral 95%: {guard['ls_uni95']*100:.3f} pp")
print(f"Margen permitido: {MARGEN_GUARDRAIL*100:.1f} pp")
print(f"p de no inferioridad: {guard['p_menor']:.6g}")


,incidencias,n,tasa_incidencia
grupo,,,
A_Control,139,8743,0.015898
B_Tratamiento,176,8743,0.020130


Cambio del guardrail B−A: 0.423 pp
Límite superior unilateral 95%: 0.754 pp
Margen permitido: 0.5 pp
p de no inferioridad: 0.351286


**Interpretación del resultado:**

El análisis del guardrail muestra que la tasa de incidencia de pago aumentó de 1,59 % en el grupo de control (A) a 2,01 % en el grupo de tratamiento (B), lo que representa una diferencia observada de +0,423 puntos porcentuales (pp). Sin embargo, el límite superior unilateral del 95 % para esta diferencia es de 0,754 pp, valor que excede el margen máximo de deterioro permitido de 0,5 pp. Dado que el límite superior supera el umbral de seguridad y el p-valor de no inferioridad (0,351) es mayor que el nivel de significancia (0,05), no existe evidencia estadística suficiente para garantizar que el guardrail esté protegido, lo cual impide recomendar el despliegue de la variante B a pesar de su mejora en conversión.

### Paso 9. Integrar la regla de decisión
| Criterio | Evidencia requerida |
|---|---|
| Diseño y calidad | Una fila por usuario, asignación persistente, muestra y seguimiento completos |
| Mejora frente a cero | p unilateral < 0,05 |
| Relevancia empresarial | Límite inferior unilateral del 95 % > +1,0 pp |
| Guardrail | Límite superior unilateral del 95 % < +0,5 pp |

La regla se fijó antes de simular. El MDE dimensiona la muestra y no reemplaza al umbral empresarial. Si no se satisface un requisito, distinguir evidencia insuficiente de evidencia de perjuicio. Cualquier experimento posterior necesita un nuevo plan; no seguir agregando datos hasta lograr significancia.


In [ ]:
cumple_estadistica = bool(conv["p_mayor"] < ALPHA)
cumple_negocio = bool(conv["li_uni95"] > UMBRAL_NEGOCIO)
cumple_guardrail = bool(guard["ls_uni95"] < MARGEN_GUARDRAIL)
cumple_diseno = bool((tam_grupos >= N_POR_GRUPO).all())
criterios = pd.Series({"Diseño validado": cumple_diseno,
    "Mejora frente a cero": cumple_estadistica,
    "Mejora superior al umbral": cumple_negocio,
    "Guardrail protegido": cumple_guardrail})
display(criterios.to_frame("cumple"))
if criterios.all():
    decision = "Recomendar B en el caso simulado, con seguimiento posterior."
else:
    decision = "No recomendar el despliegue todavía: revisar los criterios no cumplidos."
print(decision)
print("Conclusión académica con datos simulados; no valida un despliegue real.")


,cumple
Diseño validado,True
Mejora frente a cero,True
Mejora superior al umbral,True
Guardrail protegido,False


No recomendar el despliegue todavía: revisar los criterios no cumplidos.
Conclusión académica con datos simulados; no valida un despliegue real.


**Interpretación del resultado:**

El código evalúa la regla de decisión conjunta definida antes del experimento. Aunque el diseño es válido, la variante B demuestra una mejora estadísticamente significativa frente a cero y supera el umbral de relevancia empresarial de 1,0 pp, el criterio del guardrail no se cumple. El límite superior unilateral del 95 % para la diferencia en incidencias de pago (0,754 pp) excede el margen máximo de deterioro permitido (0,5 pp). Por lo tanto, la regla de decisión falla en su conjunto, llevando a la conclusión correcta de no recomendar el despliegue de la variante B todavía, ya que no se puede garantizar estadísticamente que la métrica de seguridad esté protegida.

### Pausa de decisión
| Componente | Resultado observado | Interpretación |
|---|---|---|
| p unilateral frente a cero | 2,61 × 10⁻⁸ | Existe evidencia estadística muy fuerte para rechazar H₀, indicando que la variante B mejora la conversión respecto a A. |
| Diferencia B−A en pp y lift relativo | 2,745 pp y 24,10 % | La variante B muestra una mejora descriptiva notable en la tasa de conversión respecto al grupo de control. |
| Límite inferior y umbral empresarial | 1,916 pp (umbral: 1,0 pp) | El límite inferior supera el umbral de relevancia práctica, garantizando que la mejora es empresarialmente significativa. |
| Límite superior del guardrail y margen | 0,754 pp (margen: 0,5 pp) | El límite superior excede el margen de deterioro permitido, por lo que no se puede garantizar que el guardrail esté protegido. |
| Calidad y límites del caso simulado | Muestra completa (8743 por grupo), sin datos ausentes ni contaminación. | Los datos cumplen con el diseño experimental planificado, pero al ser simulados, no validan un despliegue real sin pruebas adicionales. |

**Decisión en cuatro líneas:** recomendar B / no recomendar todavía, porque ___________. Incluya al menos dos cifras. Si falta evidencia, proponga un nuevo diseño válido; no continúe esta muestra buscando significancia.

No recomendar el despliegue todavía, porque aunque la mejora en conversión es estadísticamente significativa y supera el umbral empresarial (límite inferior de 1,916 pp > 1,0 pp), el guardrail no está protegido, ya que su límite superior (0,754 pp) excede el margen máximo de deterioro permitido (0,5 pp). Se debe investigar la causa del aumento en las incidencias de pago y diseñar un nuevo experimento que mitigue este riesgo antes de continuar.



## Actividad 7 — Retos de aplicación y retroalimentación
Se resuelven durante la clase y no generan una entrega adicional.

### Reto 1 — Estimación y umbral
Un caso presenta p = 0,02 frente a cero, mejora estimada de 0,4 pp y umbral empresarial de 3 pp. ¿Qué demuestra la prueba y qué no demuestra? ¿Qué información aporta el intervalo?

**Respuesta:** La prueba demuestra que existe una diferencia estadísticamente significativa entre las variantes (p = 0,02 < 0,05), permitiendo rechazar la hipótesis nula de que no hay efecto. Sin embargo, no demuestra que la mejora sea empresarialmente relevante, ya que la estimación puntual observada (0,4 pp) es muy inferior al umbral de negocio establecido (3 pp). El intervalo de confianza aporta el rango de valores plausibles para el efecto real en la población; si este intervalo se sitúa muy por debajo de los 3 pp, confirma que, aunque el efecto es estadísticamente real, su magnitud es demasiado pequeña para justificar la implementación desde el punto de vista práctico.


### Reto 2 — Cambio de pregunta
Si interesa cualquier diferencia entre A y B, formule H0 y H1 e indique el tipo de contraste. ¿Por qué debe elegirse antes de observar los resultados?

**Respuesta:** Si interesa detectar cualquier diferencia, las hipótesis deben formularse como H₀: p_B - p_A = 0 (no existe diferencia) y H₁: p_B - p_A ≠ 0 (existe diferencia en cualquier dirección), lo que corresponde a un contraste bilateral. Esta elección debe realizarse antes de observar los resultados para evitar el sesgo de confirmación y mantener el control de la tasa de error Tipo I; decidir el tipo de contraste a posteriori (por ejemplo, optar por unilateral solo porque la diferencia observada es positiva) infla artificialmente la significancia estadística y compromete la validez metodológica del experimento.


### Reto 3 — Guardrail no concluyente
El margen es +0,5 pp. El aumento estimado de incidencias es +0,2 pp, pero el límite superior unilateral es +0,7 pp. ¿Se ha demostrado cumplir el guardrail? ¿Es correcto afirmar que se ha demostrado perjuicio?

**Respuesta:** No se ha demostrado cumplir el guardrail, ya que la regla de no inferioridad exige que el límite superior unilateral del 95 % sea estrictamente menor que el margen permitido (+0,5 pp), condición que no se cumple al ser +0,7 pp. Tampoco es correcto afirmar que se ha demostrado perjuicio, porque el intervalo de confianza incluye valores menores a +0,5 pp (como la propia estimación puntual de +0,2 pp); esto indica simplemente que la evidencia es insuficiente o inconclusa para proteger la métrica, pero no prueba estadísticamente que el deterioro sea real.


### Reto 4 — Reproducibilidad y sesgo
Un equipo cambia la semilla hasta conseguir una decisión favorable. Explique por qué eso invalida el ejercicio. Proponga cómo documentar una exploración de escenarios sin sustituir el caso base.

**Respuesta:** Cambiar la semilla hasta conseguir una decisión favorable invalida el ejercicio porque constituye una práctica de selección de resultados (p-hacking), la cual rompe el prerregistro del experimento y hace que la tasa de error Tipo I declarada pierda su validez, transformando una prueba estadística rigurosa en una búsqueda de confirmación. Para documentar una exploración de escenarios sin sustituir el caso base, se debe mantener la semilla original fija para el análisis y la decisión oficial, reportando las variaciones (como cambios en el MDE o el baseline) en una sección aparte de "análisis de sensibilidad" o "robustez", dejando explícito que son escenarios hipotéticos y no la conclusión del experimento.



# VIII. Secuencia de trabajo para la sesión
| Momento | Minutos | Desarrollo |
|---|---:|---|
| Apertura | 10 | Pregunta orientadora y caso |
| Actividad 1 | 15 | Activación conceptual |
| Actividad 2 | 20 | Representación e hipótesis |
| Actividad 3 | 30 | Planificación, simulación y calidad |
| Receso | 15 | Pausa |
| Actividad 4 | 25 | Sensibilidad de muestra |
| Actividad 5 | 30 | Tasas, contraste e intervalos |
| Actividad 6 | 20 | Guardrail y decisión |
| Actividad 7 | 15 | Retos y ticket de salida |

Total: 180 minutos, incluido el receso.

## Ticket de salida
«Un experimento A/B no termina en el p-valor porque este solo indica la existencia de una diferencia estadística, pero no evalúa la magnitud real del efecto, su relevancia práctica para el negocio (umbral empresarial) ni el posible deterioro de métricas de seguridad (guardrails), los cuales son indispensables para tomar una decisión de implementación informada y segura».


# IX. Lista de verificación de aprendizaje
- [ ] Identifico factor, niveles y unidad experimental.
- [ ] Defino respuesta, denominador y ventana de observación.
- [ ] Formulo hipótesis antes de observar resultados.
- [ ] Distingo baseline, MDE, umbral empresarial, alfa y potencia.
- [ ] Calculo muestra antes de generar los datos del caso.
- [ ] Interpreto pp, lift, p-valor e intervalos sin confundirlos.
- [ ] Evalúo el guardrail con margen e incertidumbre.
- [ ] Distingo evidencia insuficiente de evidencia de perjuicio.
- [ ] Reconozco peeking, contaminación y selección de resultados.
- [ ] Declaro el carácter simulado y los límites del ejercicio.


# X. Referencias
- Kohavi, R., Tang, D., y Xu, Y. (2020). Trustworthy Online Controlled Experiments. Cambridge University Press.
- Bruce, P., Bruce, A., y Gedeck, P. (2020). Practical Statistics for Data Scientists, segunda edición. O’Reilly.
- statsmodels. Documentación de NormalIndPower y proportion_effectsize: https://www.statsmodels.org/stable/stats.html
- SciPy. Distribución normal, funciones cdf, sf y ppf: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html

Datos: simulación académica NexaCommerce incluida en el notebook, semilla 20260921. Los datos no proceden de las referencias bibliográficas.


# **Conclusiones**

* El desarrollo de este experimento A/B evidencia que un diseño riguroso exige definir la métrica primaria, el umbral empresarial y los guardrails antes de observar los datos, evitando así sesgos de confirmación. En el caso analizado, aunque la variante B (checkout simplificado) demostró una mejora estadísticamente significativa en la tasa de conversión (2,745 pp), superando tanto el cero como el umbral de relevancia empresarial de 1,0 pp, la evidencia no fue suficiente para recomendar su despliegue.

* Esto se debe a que el límite superior del intervalo de confianza del guardrail (incidencia de pago) alcanzó 0,754 pp, excediendo el margen máximo de deterioro permitido de 0,5 pp. En consecuencia, se concluye que la decisión de implementar una nueva variante no puede basarse únicamente en la significancia estadística de la métrica primaria, sino que requiere integrar la evidencia práctica y garantizar que las métricas de seguridad no se deterioren, validando así una decisión de negocio informada y responsable.